fairlearn/metrics/_annotated_metric_function.py

```diff
@@ -0,0 +1,60 @@
+# Copyright (c) Microsoft Corporation and Fairlearn contributors.
+# Licensed under the MIT License.
+
+from typing import Callable, Dict, List
+
+import pandas as pd
+
+
+class AnnotatedMetricFunction:
+    """Wrapper for functions, to give them args and kwargs."""
+
+    def __init__(self,
+                 *,
+                 func: Callable,
+                 args: List[str] = None,
+                 kwargs: Dict[str, str] = None):
+        self._func = func
+        self._args = ['y_true', 'y_pred']
+        if args is not None:
+            self._args = args
+        self._kwargs = dict()
+        if kwargs is not None:
+            self._kwargs = kwargs
+
+    @property
+    def func(self) -> Callable:
+        """Return the wrapped function."""
+        return self._func
+
+    @property
+    def args(self) -> List[str]:
+        """Return the list of positional arguments."""
+        return self._args
+
+    @property
+    def kwargs(self) -> Dict[str, str]:
+        """Return the mapping from column names to kwarg names."""
+        return self._kwargs
+
+    def invoke(self, df: pd.DataFrame, split_columns: Dict[str, List[str]]):
```

Comment:
do you want to have `invoke`, or make them a callable with `__call__` and have them be called as usual?

Reply:
On the whole, I prefer an `invoke()` method, although I'm not going to die on that particular hill.

Reply:
The reason I say that, is that in the place where we check what metric needs what, we can simply check

``` python
params = defaults
if hasattr(metric, "kwargs"):
    # extend params
value = metric(**params)
```
we don't need to have a separate syntax for callables and these objects there then.